packages/protocol/contracts/rrp/authorizers/Api3TokenLock.sol

```diff
@@ -0,0 +1,393 @@
+            IApi3Token(api3Token).transferFrom(
+                msg.sender,
+                address(this),
+                lockAmount
+            )
+        );
+
+        if (airnodeClient.whitelistCount == 1) {
+            IApi3RequesterRrpAuthorizer(api3RequesterRrpAuthorizer)
+                .setWhitelistExpiration(
+                _airnodeId,
+                _clientAddress,
+                type(uint64).max
+            );
+        }
+
+        emit Lock(_airnodeId, _clientAddress, msg.sender, lockAmount);
+    }
+
+    /// @notice Unlocks the API3 tokens for a given airnode-client pair.
+    /// Airnode-client pair will be unauthorized if no token lock for the pair is found.
+    /// @dev Checks whether the user (msg.sender) has already locked anything,
+    /// if the locked period has expired and if the client address is blacklisted
+    /// Transfers to msg.sender the locked amount.
+    /// @param _airnodeId The airnode id
+    /// @param _clientAddress The client address
+    function unlock(bytes32 _airnodeId, address _clientAddress)
+        external
+        override
+    {
+        AirnodeClient storage airnodeClient = tokenLocks[_airnodeId][
+            _clientAddress
+        ];
+        require(airnodeClient.lockAmountAt[msg.sender] != 0, ERROR_NOT_LOCKED);
+        require(
+            airnodeClient.canUnlockAt[msg.sender] != 0,
+            ERROR_LOCK_PERIOD_ZERO
+        );
+        require(
+            airnodeClient.canUnlockAt[msg.sender] <= block.timestamp,
+            ERROR_LOCK_PERIOD_NOT_EXPIRED
+        );
+        require(
+            !clientAddressToBlacklistStatus[_clientAddress],
+            ERROR_CLIENT_BLACKLISTED
+        );
+        uint256 amount = airnodeClient.lockAmountAt[msg.sender];
+
+        airnodeClient.lockAmountAt[msg.sender] = 0;
+        airnodeClient.canUnlockAt[msg.sender] = 0;
+        airnodeClient.whitelistCount--;
+
+        if (airnodeClient.whitelistCount == 0) {
+            IApi3RequesterRrpAuthorizer(api3RequesterRrpAuthorizer)
+                .setWhitelistExpiration(_airnodeId, _clientAddress, 0);
+        }
+
+        assert(IApi3Token(api3Token).transfer(msg.sender, amount));
+
+        emit Unlock(_airnodeId, _clientAddress, msg.sender, amount);
+    }
+
+    /// @notice User calls this when the lock amount has been decreased and wants
+    /// to withdraw the redundantly locked tokens
+    /// @param _airnodeId The airnode id
+    /// @param _clientAddress The client address
+    function withdraw(bytes32 _airnodeId, address _clientAddress)
+        external
+        override
+    {
+        require(
+            tokenLocks[_airnodeId][_clientAddress].lockAmountAt[msg.sender] >
+                lockAmount,
+            ERROR_INSUFFICIENT_AMOUNT
+        );
+        require(
+            !clientAddressToBlacklistStatus[_clientAddress],
+            ERROR_CLIENT_BLACKLISTED
+        );
+        uint256 withdrawAmount = tokenLocks[_airnodeId][_clientAddress]
```

Comment:
So there is a this statement that checks for that to avoid that. Even if this statement wasn't there, solidity subtractions now throw underflow error without using safemath.
```
require(
            tokenLocks[_airnodeId][_clientAddress].lockAmountAt[msg.sender] >
                lockAmount,
            ERROR_INSUFFICIENT_AMOUNT
);
 ```